neurots/generate/orientations.py

```diff
@@ -360,3 +435,215 @@ def compute_interval_n_tree(soma, n_trees, rng=np.random):
+                )
+
+            p = _prob(angle) / max(_prob(np.linspace(0, np.pi, 100)))
+            if with_density:
+                p *= np.sin(angle)
+            return np.clip(p, 0.0, 1.0)
+
+        return double_prob
+
+    raise ValueError(
+        f"The '{form}' value is unknown, it should be one of ['flat', 'step', 'double_step']"
+    )
+
+
+def _fit_single_3d_angles(data, neurite_type, morph_class, fit_params=None):
+    """Fit function to distribution of 3d angles for a neurite_type.
+
+    Args:
+        data (dict): bins and weights data from input_distribution
+        neurite_type (str): neurite_type to consider
+        morph_class (str): morph_class of the neuron (with_apical or without_apical)
+        fit_params (dict): specific fit parameters such as form and bounds to overwrite the defaults
+    """
+    _fit_params = deepcopy(fit_3d_angles_params)
+
+    if fit_params is not None:
+        _fit_params[morph_class][neurite_type].update(fit_params)
+
+    form = _fit_params[morph_class][neurite_type]["form"]
+    if form != "flat":
+        function = get_probability_function(form, with_density=True)
+
+        try:
+            popt = curve_fit(
+                function,
+                data["bins"],
+                data["weights"],
+                bounds=_fit_params[morph_class][neurite_type]["bounds"],
+            )[0]
+        except RuntimeError:
+            warnings.warn("Cannot fit some trunk angles, we fallback to flat distribution")
+            form = "flat"
+            popt = np.array([])
+    else:
+        popt = np.array([])
+    return {"form": form, "params": popt.tolist()}
+
+
+def _get_fit_params_from_input_parameters(parameters):
+    """Get parameter dict for fits from tmd_parameters."""
+    if "values" in parameters["orientation"] and parameters["orientation"]["values"] is not None:
+        form = parameters["orientation"]["values"].get("form")
+        bounds = parameters["orientation"]["values"].get("bounds")
+        if form is not None:
+            if bounds is None:
+                bounds = fit_3d_angles_bounds[form]
+            if parameters["orientation"]["values"].get("params") is None:
+                return {"form": form, "bounds": bounds}
+    return None
+
+
+def fit_3d_angles(tmd_parameters, tmd_distributions):
+    """Fit functions to 3d_angles from `tmd_distributions` and save in copy of `tmd_parameters`.
+
+    If if fit parameters are already in `tmd_parameters`, the fit is skipped.
+
+    We return `None` instead of copy of `tmd_parameters` if there is any `3d_angle` data that
+    was present in `tmd_distributions`, with or without fit data, so this function can be used
+    to detect if `3d_angles` modes can be used in synthesis.
+
+    Args:
+        tmd_parameters (dict): Input parameters.
+        tmd_distributions (dict): Input distributions.
+
+    Returns:
+        tmd_parmeters with fit data if 3d_angles mode is found, else None
+    """
+    with_3d = False
```

Comment:
in which case would this remain False? could this be a check implemented outside?

Reply:
if we are not in 3d_angle mode